### ISSUE TITLE
Run all tests in assembly twice with different config

### ISSUE THREAD

**Issue by LarsAndreasEk** [nunit/nunit#3270]
Hi,

I have a suite of 1000+ integration tests that I would like to run multiple times using different configurations. Concretely, the product supports two storage providers (sql + mongodb) and I would like all tests to run for each provider (both when run on CI and for all devs running the test suite locally from Visual Studio).

I can solve the CI case by running the suite twice, but that would not suffice for the developers. Another option is to use parameterized test fixtures, but I don't really want to change all the tests.

Any tips or tricks that might help?

Br
Lars Andreas

**Comment by mlidbom** [comment]
Bumping this one since I have exactly the same need except that for me the storage providers are MySql, Postgre, SQL Server, and Oracle. Any tips would be greatly appreciated.

**Comment by ChrisMaddock** [comment]
How are you running your tests? If it's with the NUnit console, you might want to look at setting up 4 different config files, and specifying the different configs in an [NUnit Project FIle](https://github.com/nunit/docs/wiki/NUnit-Test-Projects).

**Comment by mlidbom** [comment]
Thanks for suggestion. If I was using the console that sounds like it would do the trick. 
I'm using NCrunch ( https://www.ncrunch.net)  to run my tests though and I find it to be indispensable to be able to run my tests that way. I could not imagine giving that workflow up.

**Comment by ChrisMaddock** [comment]
Ah ok. I don't know myself how NCrunch integrates with NUnit - you may be better going to their support instead. Does it have any support for `TestParameters`? That would be another option.

**Comment by mlidbom** [comment]
I'm unaware of any NUnit feature not supported by NCrunch.
But I thought TestParameters were targeted to a single test or test fixture. 
I have many hundreds of tests and I would really like to find a solution that does not require me to perform open heart surgery on them en masse. 
Currently I am, reluctantly, trying the option of using a shared project ( shared source, not assembly: https://docs.microsoft.com/en-us/xamarin/cross-platform/app-fundamentals/shared-projects?tabs=windows ) that contains all the test code and then creating a project per storage provider that references that shared project. These projects would be virtually empty, just contain something that modifies the environment so that the shared test code knows which provider to use.

**Comment by mlidbom** [comment]
This seems to work, but the multiplied virtually empty projects really are not making me happy. Neither is the additional build time. NCrunch having to build the entire test project once per provider does nothing good for the number of seconds it takes me to know if I broke something with the last change.

**Comment by ChrisMaddock** [comment]
No, TestParameters are global within an assembly. (`TestContext.TestParameters` are the ones I'm referring to?) They're set in different ways depending on your test runner - the .runsettings file for the NUnit VS adapter, the -testparams flag for the NUnit Console etc. As before, I'm don't know how NCrunch works as to how you'd be able set them. They're a second feature that could potentially be used to achieve this, with the right options in a runner.

Those and the NUnit project files are the only two features I can think of that would support this.

**Comment by mlidbom** [comment]
Thanks again @ChrisMaddock. I'll look into TestContext.TestParameters and how/whether NCrunch supports them. If not I guess I'll go with a shared project for now. I started exploring it after posting my first comment here, I'm just about done getting it working, and the only real downsides are the ones I gripe about above.  so I guess the pragmatic choice might be to go for that solution unless/until I run into any major problems with it.

**Comment by mlidbom** [comment]
My problem has been solved cleanly by an NCrunch feature that is currently in testing: the DistributeByCapabilitiesAttribute: https://forum.ncrunch.net/yaf_postsm14815_Find-out-which-DistributeByCapabilitiesAttribute-capability-a-test-is-executing-under.aspx

I'm now running that version and it seems to do exactly what I need. Posting that here for anyone else that might find there way here through google the way that I did.

**Comment by ChrisMaddock** [comment]
Nice, cool feature! 😊
